healthcheck/core/src/main/java/org/apache/felix/hc/core/impl/executor/async/cron/CronScheduler.java

```diff
@@ -0,0 +1,170 @@
+package org.apache.felix.hc.core.impl.executor.async.cron;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class CronScheduler implements AutoCloseable {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Object used for internal locking/notifications.
+     */
+    private final Object monitor = new Object();
+
+    /**
+     * List of managed tasks.
+     */
+    private final Map<String, CronSchedulerTask> tasks = new HashMap<>();
+
+    /**
+     * Task executor instance.
+     */
+    public final ExecutorService tasksExecutor;
+
+    /**
+     * Thread that makes all scheduling job.
+     */
+    public final SchedulerThread schedulerThread;
+
+    /**
+     * If scheduler is active or not.
+     */
+    private boolean isActive;
```

Comment:
Should `isActive` be `volatile` ? 

The access to it is not synchronized, so changes between `shutdown`, `isShutdown` and `run` might not be visible in other threads.

Reply:
Fixed